src/bundles/ipfs-provider.js

```diff
@@ -0,0 +1,323 @@
+    }
+    case ACTIONS.IPFS_CONNECT_SUCCEED: {
+      return { ...state, failed: false }
+    }
+    case ACTIONS.IPFS_CONNECT_FAILED: {
+      return { ...state, failed: true }
+    }
+    default: {
+      return state
+    }
+  }
+}
+
+/**
+ * @returns {Model}
+ */
+const init = () => {
+  return {
+    apiAddress: readAPIAddressSetting(),
+    provider: null,
+    failed: false,
+    ready: false,
+    invalidAddress: false,
+    pendingFirstConnection: false
+  }
+}
+
+const readAPIAddressSetting = () => {
+  const setting = readSetting('ipfsApi')
+  return setting == null ? null : asAPIOptions(setting)
+}
+
+const asAPIOptions = (value) => asHttpClientOptions(value) || asMultiaddress(value) || asURL(value)
+
+const asURL = (value) => {
+  try {
+    return new URL(value).toString()
+  } catch (_) {
+    return null
+  }
+}
+
+const asMultiaddress = (value) => {
+  // ignore empty string, as it will produce '/'
+  if (value != null && value !== '') {
+    try {
+      return multiaddr(value).toString()
+    } catch (_) {}
+  }
+  return null
+}
+
+const asHttpClientOptions = (value) =>
+  typeof value === 'string' ? parseHTTPClientOptions(value) : readHTTPClinetOptions(value)
+
+const parseHTTPClientOptions = (input) => {
+  // Try parsing and reading as json
+  try {
+    return readHTTPClinetOptions(JSON.parse(input))
+  } catch (_) {}
+
+  // turn URL with inlined basic auth into client options object
+  try {
+    const uri = new URL(input)
+    const { username, password } = uri
+    if (username && password) {
+      return {
+        host: uri.hostname,
+        port: uri.port || (uri.protocol === 'https:' ? '443' : '80'),
+        protocol: uri.protocol.slice(0, -1), // trim out ':' at the end
+        apiPath: (uri.pathname !== '/' ? uri.pathname : 'api/v0'),
+        headers: {
+          authorization: `Basic ${btoa(username + ':' + password)}`
+        }
+      }
+    }
+  } catch (_) { }
+
+  return null
+}
+
+const readHTTPClinetOptions = (value) => {
+  // https://github.com/ipfs/js-ipfs/tree/master/packages/ipfs-http-client#importing-the-module-and-usage
+  if (value && (value.host || value.apiPath || value.protocol || value.port || value.headers)) {
+    return value
+  } else {
+    return null
+  }
+}
+
+const readSetting = (id) => {
+  /** @type {string|null} */
+  let setting = null
+  if (window.localStorage) {
+    try {
+      setting = window.localStorage.getItem(id)
+    } catch (error) {
+      console.error(`Error reading '${id}' value from localStorage`, error)
+    }
+
+    try {
+      return JSON.parse(setting || '')
+    } catch (_) {
+      // res was probably a string, so pass it on.
+      return setting
+    }
+  }
+
+  return setting
+}
+
+const writeSetting = (id, value) => {
+  try {
+    window.localStorage.setItem(id, JSON.stringify(value))
+  } catch (error) {
+    console.error(`Error writing '${id}' value to localStorage`, error)
+  }
+}
```

Comment:
There  is no UI for changing HTTP API endpoint  (and won't be after switch to js-ipfs)
```suggestion
```